telmos_main.py

```diff
@@ -36,56 +24,253 @@ def telmos_main(delta_root, tmfs_root, tel_year, tel_id, tel_scenario,
+    sw_attr["5"] = sw_array[4,:,5].sum()
+    sw_attr["4"] = sw_array[3,:,5].sum()
+    sw_attr["14"] = sw_array[7,:,5].sum()
+    sw_attr["17"] = sw_array[8,:,5].sum()
+
+    sw_array[0,:,5] *= (sw_prod["1"] / sw_attr["1"])
+    sw_array[4,:,5] *= (sw_prod["5"] / sw_attr["5"])
+    sw_array[3,:,5] *= (sw_prod["4"] / sw_attr["4"])
+    sw_array[7,:,5] *= (sw_prod["14"] / sw_attr["14"])
+    sw_array[8,:,5] *= (sw_prod["17"] / sw_attr["17"])
+    
+    sw_cte_array = np.zeros_like(cte_data, dtype="float")
+    prod_col_idxs = np.array([1,2,3,5,6,7,4])
+    attr_growth_idxs = [None, 2, 1, None, None, 2, 1, None, None]
+    for j in range(sw_cte_array.shape[0]):
+        if j < 8:
+            sw_cte_array[j,:,1:8] = (cte_data[j,:,1:8] * 
+                        production_growth[:,prod_col_idxs+(j*8)] *
+                        airport_growth[:,None])
+        else:
+            sw_cte_array[j,:,1:8] = (cte_data[j,:,1:8] * 
+                        production_growth[:,prod_col_idxs+((j-1)*8)] *
+                        airport_growth[:,None])
+        # These columns do the following
+        if j in [0,3,4,7,8]:
+            sw_cte_array[j,:,8] = sw_array[j,:,5]
+        # Otherwise...
+        else:
+            sw_cte_array[j,:,8] = (cte_data[j,:,8] * 
+                        production_growth[:,attr_growth_idxs[j]] * 
+                        airport_growth)
+    
+    return (sw_array, sw_cte_array)
+
+def save_trip_end_files(file_names, trip_ends, base_path, precision):
+    for i, t_file in zip(range(trip_ends.shape[0]), file_names):
+        path = os.path.join(base_path, t_file.replace("_ALL", ""))
+        np.savetxt(path, np.concatenate(
+                (np.arange(trip_ends[i].shape[0])[:,None]+1, trip_ends[i][:,1:]),axis=1),
+                    delimiter=", ", 
+                    fmt=["%d"]+["%." + str(precision) + "f" for x in range(trip_ends[i].shape[1]-1)])
+                
+
+def telmos_main(delta_root, tmfs_root, tel_year, tel_id, tel_scenario,
+                base_year, base_id, base_scenario, is_rebasing_run=True,
+                log_func=print, just_pivots=False):
+    '''
+    Conversion of the TMfS14 Visual Basic application 
+    '''
+    
+    
+    # Read in the trip rate matrices into multi-dim array
+    factors_base = os.path.join(tmfs_root, "Factors")
+    p_trip_rate_array = read_trip_rates(factors_base, just_pivots)
     
     # Load in the student factors and attraction factors separately
-    filenames = ["Student Factors.txt", "Attraction Factors.txt"]
-    student_factors = np.loadtxt(os.path.join(factors_base, filenames[0]))
-    attraction_factors = np.loadtxt(os.path.join(factors_base, filenames[1]))
+    attraction_file = "Attraction Factors.txt"
+    attraction_factors = np.loadtxt(os.path.join(factors_base, attraction_file))
+        
+    # [Male student factor, Female student factor]
+    # Used to factor the productions down
+    student_factors = np.array([0.2794, 0.2453])
```

Comment:
These shouldn't be an array. There's only two values, so having them in an array just makes their actual meaning more opaque.

I think one option might be to make them global factors that indicate what they're used for, at the top of the file, e.g.

```python
import numpy as np
import pandas as pd


MALE_STUDENT_FACTOR = 0.2794
FEMALE_STUDENT_FACTOR = 0.2453

def read_trip_rates(factors_dir, just_pivots):
# ...
```